src/main/kotlin/org/rust/toml/CargoCrateDocLineMarkerProvider.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml
+
+import com.intellij.codeHighlighting.Pass
+import com.intellij.codeInsight.daemon.LineMarkerInfo
+import com.intellij.codeInsight.daemon.LineMarkerProvider
+import com.intellij.ide.BrowserUtil
+import com.intellij.openapi.editor.markup.GutterIconRenderer
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil;
+import org.rust.ide.icons.RsIcons
+import org.toml.lang.psi.TomlInlineTable
+import org.toml.lang.psi.TomlKeyValue
+import org.toml.lang.psi.TomlLiteral
+import org.toml.lang.psi.TomlTable
+
+class CargoCrateDocLineMarkerProvider : LineMarkerProvider {
+    override fun getLineMarkerInfo(element: PsiElement) = null
+
+    override fun collectSlowLineMarkers(elements: MutableList<PsiElement>, result: MutableCollection<LineMarkerInfo<PsiElement>>) {
+        if (!tomlPluginIsAbiCompatible()) return
+        for (el in elements) {
+            val file = el.containingFile
+            if (file.name.toLowerCase() != "cargo.toml") continue
+            if (el !is TomlTable) continue
+            result += annotateTable(el)
+        }
+    }
+
+    private fun annotateTable(el: TomlTable): Collection<LineMarkerInfo<PsiElement>> {
+        val names = el.header.names.map { it.text }
+        val test = names.getOrNull(names.size - 2)
+        val lastName = names.lastOrNull() ?: return emptyList()
+        if (test != null && (test == "dependencies" || test == "dev-dependencies" || test == "build-dependencies")) {
+            val version = el.entries.find { it.name == "version" }?.value?.text?.trimVersion
+            val lineMarkerInfo = genLineMarkerInfo(el.header, lastName, version) ?: return emptyList()
+            return listOf(lineMarkerInfo)
+        }
+        if (lastName != "dependencies" && lastName != "dev-dependencies" && lastName != "build-dependencies") return emptyList()
+        return el.entries.mapNotNull {
+            val pkgName = it.name
+            val pkgVersion = it.version
+            genLineMarkerInfo(it, pkgName, pkgVersion)
+        }
+    }
+
+    private fun genLineMarkerInfo(element: PsiElement, name: String, version: String?): LineMarkerInfo<PsiElement>? {
+        val version = version ?: return null
+        val anchor = PsiTreeUtil.getDeepestFirst(element)
```

Comment:
Hm, it's super non-obvious what element is returned here... Could we be somehow more specific?

Reply:
I change this to fix the problem with the leaf.

Reply:
Is this okay?

Reply:
I think it's better to anchor on the specific toml element, instead of using just `PsiElement`. So, we probably want to annotate `TomlKey` child with type `BARE_KEY`:

![image](https://user-images.githubusercontent.com/1711539/33491528-9c63c5e0-d6cb-11e7-82a0-7ba0ce72dbc6.png)

Reply:
better?

Reply:
Most definitely!